arangod/Agency/TransactionBuilder.h

```diff
@@ -32,210 +32,214 @@
-  precs_trx isEmpty(K&& k) && {
-    _buffer.setValue(std::forward<K>(k));
-    _buffer.openObject();
-    _buffer.setKey("oldEmpty", true);
-    _buffer.closeObject();
-    return precs_trx(std::move(_buffer));
-  }
-
-  T done() && {
-    _buffer.closeObject();
-    _buffer.setValue(VPackValue(AgencyWriteTransaction::randomClientId()));
-    _buffer.closeArray();
-    return T(std::move(_buffer));
-  }
-
- private:
-  friend write_trx<B, T>;
-  friend T;
-
-  precs_trx(buffer &&b) : _buffer(std::move(b)) {}
-  buffer _buffer;
-};
-
-template<typename B, typename T>
-struct write_trx {
-  using buffer = buffer_mapper<B>;
-
-  template<typename K, typename V>
-  write_trx set(K&& k, V&& v) && {
-    _buffer.setValue(std::forward<K>(k));
-    _buffer.openObject();
-    _buffer.setKey("op", "set");
-    _buffer.setKey("new", std::forward<V>(v));
-    _buffer.closeObject();
-    return write_trx(std::move(_buffer));
-  }
-
-  template<typename K, typename F>
-  write_trx emplace(K&& k, F f) && {
-    _buffer.setValue(std::forward<K>(k));
-    _buffer.openObject();
-    f(_buffer.userObject());
-    _buffer.closeObject();
-    return write_trx(std::move(_buffer));
-  }
-
-  template<typename K>
-  write_trx remove(K&& k) && {
-    _buffer.setValue(std::forward<K>(k));
-    _buffer.openObject();
-    _buffer.setKey("op", "delete");
-    _buffer.closeObject();
-    return write_trx(std::move(_buffer));
-  }
-
-  template<typename K>
-  write_trx inc(K&& k, uint64_t delta = 1) && {
-    _buffer.setValue(std::forward<K>(k));
-    _buffer.openObject();
-    _buffer.setKey("op", "increment");
-    _buffer.setKey("delta", delta);
-    _buffer.closeObject();
-    return write_trx(std::move(_buffer));
-  }
-
-  T done() && { _buffer.closeObject();
-    _buffer.setValue(VPackSlice::emptyObjectSlice());
-    _buffer.setValue(VPackValue(AgencyWriteTransaction::randomClientId()));
-    _buffer.closeArray();
-    return T(std::move(_buffer));
-  }
-
-  precs_trx<B, T> precs() && {
-    _buffer.closeObject();
-    _buffer.openObject();
-    return std::move(_buffer);
-  }
-  write_trx& operator=(write_trx&&) = default;
-
- private:
-  friend T;
-
-  explicit write_trx(buffer &&b) : _buffer(std::move(b)) {}
-  buffer _buffer;
-};
-
+namespace detail {
+template <typename T>
+struct moving_ptr {
+  moving_ptr(T* p) : _ptr(p) {}
+  ~moving_ptr() = default;
 
-template<typename B>
-struct envelope {
-  using buffer = buffer_mapper<B>;
+  moving_ptr& operator=(moving_ptr const&) = delete;
+  moving_ptr(moving_ptr const&) = delete;
 
-  read_trx<B, envelope> read() && {
-    _buffer.openArray();
-    return read_trx<B, envelope>(std::move(_buffer));
-  }
+  moving_ptr& operator=(moving_ptr&& o) {
+    reset(o.release());
+    return *this;
+  };
+  moving_ptr(moving_ptr&& o) { *this = std::move(o); }
 
-  write_trx<B, envelope> write() && {
-    _buffer.openArray();
-    _buffer.openObject();
-    return write_trx<B, envelope>(std::move(_buffer));
+  T* release() {
```

Comment:
This will be called from destructors and such, so would be good to mark this as being `noexcept`.